extensions/wikia/WikiFactory/SpecialWikiFactory_body.php

```diff
@@ -61,160 +65,201 @@ public function execute( $subpage ) {
 			$wgOut->addHTML( $this->longStats( $lang ) );
-		}
-		elseif ( strtolower($subpage) === "add.variable" ) {
-			$varOverrides = array();
-			$wgOut->addHTML( $this->doAddVariableForm($varOverrides) ); // process the post (if relevant).
-			$wgOut->addHTML( $this->addVariableForm($varOverrides) ); // display the form
-		}
-		else {
+		} elseif ( strtolower( $subpage ) === "add.variable" ) {
+			$varOverrides = [];
+			$wgOut->addHTML( $this->doAddVariableForm( $varOverrides ) ); // process the post (if relevant).
+			$wgOut->addHTML( $this->addVariableForm( $varOverrides ) ); // display the form
+		} else {
 			$subpage = ( $subpage == "/" ) ? null : $subpage;
-			$oWiki = $this->getWikiData( $subpage );
+			$this->handleNavigationTab( $subpage );
+			$cityId = $this->findCityId( $subpage );
 
-			if( !isset( $oWiki->city_id )) {
+			if ( empty( $cityId ) ) {
 				$this->doWikiSelector();
+				return;
 			}
-			else {
-				$this->mWiki = $oWiki;
-				$this->mTags = new WikiFactoryTags( $this->mWiki->city_id );
-				$this->doWikiForm( );
+
+			$this->mTitle = Title::makeTitle(
+				NS_SPECIAL,
+				implode( '/', [ 'WikiFactory', $cityId, $this->mTab ] )
+			);
+
+			if ( !preg_match("/$cityId/", $subpage ) ) {
+				$wgOut->redirect( $this->mTitle->getLocalURL() );
+				return;
 			}
+
+			$wiki = $this->getWikiData( $subpage );
+
+			$this->mWiki = $wiki;
+			$this->mTags = new WikiFactoryTags( $this->mWiki->city_id );
+
+			$this->doWikiForm();
 		}
 	}
 
 	/**
 	 * getWikiData
 	 *
-	 * use subpage as param and try to find wiki which match criteria
-	 *
-	 * @access private
+	 * Use subpage as param and try to find wiki which match criteria
 	 *
 	 * @param mixed $subpage:
 	 *
 	 * @return mixed: database row from city_list
 	 */
 	private function getWikiData( $subpage ) {
-		global $wgRequest, $wgWikiaBaseDomain;
+		$cityId = $this->findCityId( $subpage );
+		if ( empty( $cityId ) ) {
+			return null;
+		}
 
-		$domain	= $wgRequest->getVal( "wpCityDomain", null );
-		$cityid	= $wgRequest->getVal( "cityid", null );
-		$tab = "variables";
-		if( is_null( $cityid ) && ( isset( $subpage ) || isset( $domain ) ) ) {
+		return WikiFactory::getWikiByID( $cityId );
+	}
 
-			/**
-			 * if there is # in subpage we are switching tab
-			 */
-			if( strpos( $subpage, "/" ) ) {
-				$parts = explode( "/", $subpage, 3 );
-				if( is_array( $parts ) && sizeof( $parts ) >= 2 ) {
-					$subpage = $parts[0];
-					$tab = $parts[1];
-					if ( ( $tab === "variables" ) && ( isset($parts[2]) ) ) {
-						$this->mVariableName = trim($parts[2]);
-					}
-					if ( ( $tab === "tags" || $tab === "findtags"  ) && ( isset($parts[2]) ) ) {
-						$this->mSearchTag = trim($parts[2]);
-					}
-				}
-			}
+	/**
+	 * Take care of setting internal variables necessary to handle the different
+	 * navigation tabs we could be on
+	 *
+	 * @param string $subpage
+	 */
+	private function handleNavigationTab( $subpage ) {
+		// The default
+		$this->mTab = "variables";
```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Define a constant instead of duplicating this literal "variables" 3 times. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS1192)